### ISSUE TITLE
Better projectile cache

### ISSUE THREAD

**Issue by jasonkuhrt** [syl20bnr/spacemacs#848]
After a filesystem change:

- [ ] `projectile` cache is invalid
- [ ] `projectile` has invalid cache even after `emacs` restart

`projectile` should be setup to update on any file system change. The fact that users have to think about this is like going back to 1970s or something.

Currently I have to manually do `spc p I`. Is this is a limitation of `projectile` or how spacemacs has it configured?

**Comment by syl20bnr** [comment]
This is how projectile works. Since it is supposed to be cross-platform I guess it is not trivial to implement an auto-update of the cache. There is some ways to keep it in sync with git operation though, I did not dig it though.
Invalidating the cache can be expensive so it is up to the user to setup the hook to perform automatic update of the cache.
Summoning @tuhdo if he has some cool tricks to share for this particular case :-)

**Comment by jasonkuhrt** [comment]
I feel like some opt-in systems should be sourced from the community for common platforms `linux` / `windows` / `osx` etc. I only care about `osx` currently but yes honour the cross-platform spirit.

**Comment by tuhdo** [comment]
Yes, it's possible to implement cross-platform cache auto-update with [file-notify](https://www.gnu.org/software/emacs/manual/html_node/elisp/File-Notifications.html) in Emacs. Here is an example:

```elisp
(defun my-notify-callback (event)
  (message "Event %S" event))

(mapc (lambda (d)
        (file-notify-add-watch
         d '(change attribute-change) 'my-notify-callback))
      (projectile-current-project-dirs))
```

After executing the above, changes made to all project directories will be monitored and reported. For example, you create a new file with `touch`, Emacs will print a message saying a new file is created. Currently, Projectile already has function for adding/removing files in current cache. We only need to handle notified events (create and delete) to update the file cache properly and it's done. I will implement and push it to Projectile at some point.

**Comment by syl20bnr** [comment]
@tuhdo :heart:  :+1:

**Comment by StreakyCobra** [comment]
This has probably been fixed by #3802. Closing.
